### ISSUE TITLE
How to use TinySoftwareSPI?

### ISSUE THREAD

**Issue by Krb686** [TCWORLD/ATTinyCore#8]
I've been looking for a way to program an ATtiny85 using an ATmega328, but being able to select which pins are SS, MOSI, MISO because I can't use the normal SPI pins.  Some googling led to your code (TinySoftwareSPI), but I don't understand how I would use it.  Also since your email is hidden, had to try contacting through issue report.  Sorry about that but that seems to be the only way

**Comment by TCWORLD** [comment]
For programming an ATTiny85, you use an Serial Programmer. For this you have to use the normal SPI pins, you don't have a choice what pins to use - there is a physical hardware interface built into the IC). TinySoftwareSPI won't help you.

There is an alternative if you are willing to sacrifice 512bytes of Flash and that is to install a bootloader. ATTinyCore comes with a bootloader for the Attiny85. You have to use the SPI pins to program in the bootloader the first time, but after that you only require two pins for programming - RX and TX - which connect to a standard USB-Serial port.
The pins used for this are built into the bootloader, but it would be possible to recompile the bootloader to use different pins of your choosing - if you let me know what pins you have free, I will compile a bootloader hex file for you which uses those pins.

**Comment by Krb686** [comment]
I appreciate the quick response.  My problem isn't exactly what I mentioned
before, I just simplified it because I figured I could apply your answer to
my actual problem.  In reality, I wanted to try to use an ATmega328 to
program multiple chips using a ZIF socket.  Specifically, I wanted to be
able to program the ATtiny84, 85, 2313, and 328.  Since each of those chips
obviously has a different form factor, I would need to have MISO, MOSI, and
SCK in multiple places, and I wanted to do it without a switching solution.

I originally created a question on StackExchange for this topic, with the
intent to use a switching solution, where several people told me it would
be easier to do it entirely in software.  That question is here:

http://electronics.stackexchange.com/questions/87260/attiny-atmega-programmer-board-need-a-6-x-13-mux-solution

**Comment by Krb686** [comment]
They suggested using "software SPI".  As I understand it, SPI can be done
from any digital pin, also known as "bit banging"? Isn't that what your
TinySoftwareSPI library does? Why then would I not be able to connect any 3
digital pins, not necessarily the dedicated MISO, MOSI, and SCK, and
program another chip using those?

**Comment by TCWORLD** [comment]
I see what you mean now. Yes you could do that with TinySoftwareSPI. The clock frequency of SCLK is about 250kHz, which cannot be changed. 

TinySoftwareSPI has the same syntax as the Arduino SPI library with the following functions:

SPI1.transfer(byte); //transmits 'byte' and returns any received 'byte'
SPI1.setDataMode(mode); //The SPI mode, e.g. SPI_MODE0
SPI1.setBitOrder(order); //Set transmission bit order to MSBFIRST or LSBFIRST

**Comment by pfriedel** [comment]
Well, are you suggesting using software SPI on the programming device or on the device being programmed?  The latter is a catch 22.  Software SPI only works once the chip is up and running code, but for the initial programming of a blank chip you have to use the hardware SPI pins.  The tiny85/2313 can share a programmer easily enough (jumper pin 4 for gnd on the 8 pin chips, pull it for the 14 pin chips), but I think the 328 needs too much supplemental wiring (vcc, gnd, crystal, etc) to be very effective with an all-in-one ZIF solution without hanging jumpers all around it.

**Comment by Krb686** [comment]
Yeah I should have clarified, I meant using soft SPI on the programmer, not
the chip being programmed. The programmer would be a pre-loaded 328.  As
far as VCC and GND, I was thinking of using several digital pins from the
328 to control some tri-state logic buffers, another solution suggested to
me on SE.  I just recently got my 328 programmed using the "minimal" way
without crystal or capacitors, so I don't think those should be an issue.
 Does this all still seem possible?

In a nutshell, here's where the different pin locations are for each chip.

http://i306.photobucket.com/albums/nn255/Krb686/zif_pins.png

Each color is a chip type and the black is the socket.  So it amounts to
what, 15 locations? So I need 15 digital pins from the 328?

edit: Well there are 11 different signal locations, and 7 different VCC/GND locations, so 18 pins?

**Comment by TCWORLD** [comment]
Actually that makes it easier. If you are only programming one chip at any given time you can do:

  #include "TinySoftwareSPI.h"
  ...
  ...
  //Something to select the chip
  ...
  //Load the correct pins from some lookup table
  byte CLOCK_Pin = ClockPins[whichMCU];
  byte MOSI_Pin = MOSIPins[whichMCU];
  byte MISO_Pin = MISOPins[whichMCU];
  byte SS_Pin = Resets[whichMCU];
  //Start the SPI on the correct pins:
  SPI.begin(CLOCK_Pin, MOSI_Pin, MISO_Pin, SS_Pin);

//Do the programming. You can use the Arduino ISP code as that uses the SPI library for which my library is a drop in software replacement

**Comment by Krb686** [comment]
Thanks a lot for the help and for this cool library!

**Comment by Krb686** [comment]
This is odd, I guess they never took the time to update the ArduinoISP
sketch even in the latest version of the IDE (1.0.5).  My ArduinoISP sketch
most definitely does not use the SPI library, instead it is the old sketch
that writes directly to the registers.  Even the version hosted by adafruit
doesn't make use of the SPI library.

I did manage to find a version that uses the SPI library on Randall Bohn's
github, but even there he failed to update the version as that one says
version "04m3" as well as my register version too.

**Comment by TCWORLD** [comment]
There is a new one here which uses SPI.h:
https://github.com/rsbohn/ArduinoISP/blob/master/ArduinoISP/ArduinoISP.ino

**Comment by TCWORLD** [comment]
I'm closing this due to inactivity. If you still have a problem, reopen it.
